packages/admin-graphql-api-utilities/src/index.ts

```diff
@@ -10,10 +27,60 @@ export function parseGid(gid: string): string {
   throw new Error(`Invalid gid: ${gid}`);
 }
 
-export function composeGid(key: string, id: number | string): string {
-  return `gid://shopify/${key}/${id}`;
+export function parseGidWithParams(gid: string): ParsedGID {
+  // appends forward slash to help identify invalid id
+  const id = `/${gid}`;
+  const matches = GID_REGEXP.exec(id);
+  if (matches && matches[1] !== undefined) {
+    const params =
+      matches[2] === undefined
+        ? {}
+        : matches[2]
+            .split('&')
+            .reduce<Record<string, string>>((acc, keyVal) => {
+              const [key, value] = keyVal.split('=');
+              acc[key] = value;
+              return acc;
+            }, {});
+
+    return {
+      id: matches[1],
+      params,
+    };
+  }
+  throw new Error(`Invalid gid: ${gid}`);
 }
 
+export function createComposeGidFunction(namescape: string) {
+  return function composeGid(
+    key: string,
+    id: number | string,
+    params: Record<string, string> = {},
+  ): string {
+    const gid = `gid://${namescape}/${key}/${id}`;
+    const paramKeys = Object.keys(params);
+
+    if (paramKeys.length === 0) {
+      return gid;
+    }
+
+    const queryParams = paramKeys
```

Comment:
Do we accept an array of query params? Something like `foo[]=hello&foo[]=world` if so, I think this will not work. Maybe it's overkill 🤷🏼‍♂️ Should we use URLSearchParams instead also here?

Reply:
Yeah, didn't want to support arrays at first because I don't think we should really be over-using these params 😅

I believe we could go with `URLSearchParams` to parse/compose the params, but not with the full URL object because it seems we should still support unprefixed gid for some reasons. Not so sure why we would even have unprefixed gids but it's in the tests lol